### PR TITLE
expose AppendBackward of ProgramDesc to python

### PR DESCRIPTION
**Pull request by jacquesqiao**
fix: https://github.com/PaddlePaddle/Paddle/issues/4694

### CODE REVIEW

File: paddle/framework/backward.h [hunk 1]
```diff
@@ -27,6 +27,8 @@ extern std::unique_ptr<OperatorBase> Backward(
     const OperatorBase& forwardOp,
     const std::unordered_set<std::string>& no_grad_vars);
 
+// TODO(someone): Add target as parameter and generate backward op
```
Comment (wangkuiyi) [inline paddle/framework/backward.h]: Let's assign this TODO to some specific person when we create it.

Comment (jacquesqiao) [reply paddle/framework/backward.h]: done

File: paddle/pybind/protobuf.cc [hunk 2]
```diff
@@ -116,6 +117,11 @@ void BindProgramDesc(py::module &m) {
                   py::return_value_policy::reference)
       .def("append_block", &ProgramDescBind::AppendBlock,
            py::return_value_policy::reference)
+      .def("backward",
```
Comment (Canpio) [inline paddle/pybind/protobuf.cc]: I think `append_backward` is a better name. `backward` means the process that executes gradient op one by one, not append them.

File: python/paddle/v2/framework/tests/test_program.py [hunk 3]
```diff
@@ -31,6 +33,21 @@ def test_program(self):
...
+        self.assertIsNotNone(block)
+
+        sum_op_desc = block.append_op()
+        sum_op_desc.set_type("sum")
+        sum_op_desc.set_input("X", ["x1", "x2"])
+        sum_op_desc.set_output("Out", ["out"])
+
+        self.assertEqual(len(block.all_ops()), 1)
+        prog.backward(set())
+        self.assertEqual(len(block.all_ops()), 3)
```
Comment (Canpio) [inline python/paddle/v2/framework/tests/test_program.py]: Add more check here to make sure additional ops are exactly what we want.

Comment (jacquesqiao) [reply python/paddle/v2/framework/tests/test_program.py]: ok

File: paddle/pybind/protobuf.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/pybind/protobuf.cc]: done

File: python/paddle/v2/framework/tests/test_program.py (hunk 3 above)
Comment (jacquesqiao) [reply python/paddle/v2/framework/tests/test_program.py]: done

**Pull request merged**